hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/opt/physical/index/IndexResolver.java

```diff
@@ -135,21 +159,122 @@ private IndexResolver() {
+
+            if (relAscending != null) {
+                rels.add(relAscending);
+            }
+
+            if (relDescending != null) {
+                rels.add(relDescending);
             }
         }
 
         return rels;
     }
 
+
+    /**
+     * Filters out index scans which collation is covered (prefix based) by another index scan in the rels.
+     *
+     * @param rels the list of index scans
+     * @return the list of filtered out index scans
+     */
+    @SuppressWarnings("checkstyle:NPathComplexity")
+    private static List<RelNode> excludeCoveredCollations(List<RelNode> rels) {
+        // Order the index scans based on their collation
+        TreeMap<RelCollation, RelNode> relsTreeMap = new TreeMap<>((coll1, coll2) -> {
+            // Compare the collations field by field
+            int coll1Size = coll1.getFieldCollations().size();
+            int coll2Size = coll2.getFieldCollations().size();
+
+            for (int i = 0; i < coll1Size; ++i) {
+                if (i >= coll2Size) {
+                    // The coll1 has more fields and the prefixes are equal
+                    return 1;
+                }
+
+                RelFieldCollation fieldColl1 = coll1.getFieldCollations().get(i);
+                RelFieldCollation fieldColl2 = coll2.getFieldCollations().get(i);
+                // First, compare directions
+                int cmp = fieldColl1.getDirection().compareTo(fieldColl2.getDirection());
+                if (cmp == 0) {
+                    // Directions are the same
+                    if (fieldColl1.getFieldIndex() == fieldColl2.getFieldIndex()) {
+                        // And fieldIndex is the same, try the next field
+                        continue;
+                    } else {
+                        return Integer.compare(fieldColl1.getFieldIndex(), fieldColl2.getFieldIndex());
+                    }
+                }
+                return cmp;
+            }
+
+            // All the fields from coll1 are equal to the fields from coll2, compare the size
+            return Integer.compare(coll1Size, coll2Size);
+        });
+
+        // Put the rels into the ordered TreeMap
+        for (RelNode rel : rels) {
+            relsTreeMap.put(rel.getTraitSet().getTrait(RelCollationTraitDef.INSTANCE), rel);
+        }
+
+        List<RelNode> resultRels = new ArrayList<>();
+        Map.Entry<RelCollation, RelNode> prevEntry = null;
+        // Go through the ordered collations and exclude covered ones
+        for (Map.Entry<RelCollation, RelNode> entry : relsTreeMap.descendingMap().entrySet()) {
+            RelCollation collation = entry.getKey();
+            RelNode relNode = entry.getValue();
+
+            if (prevEntry == null) {
+                resultRels.add(relNode);
+                prevEntry = entry;
+            } else {
+                RelCollation prevCollation = prevEntry.getKey();
```

Comment:
We cannot throw out covered collations if there is an index scan found. Consider that we resolved the following candidates:
1. [a ASC, b ASC] -> index scan
2. [an ASC] -> index lookup
The second relation is covered by the first one, but it has a filter, whilst the first one does the full scan. Therefore, we may loose a very good plan with a lookup.

Reply:
Addressed in the https://github.com/hazelcast/hazelcast/pull/18013/commits/30dc043f41e9d68727328ffc0302413f07d7bb2e the full scan never "covers" the full scans.